lib/openshift4-operators.libsonnet

```diff
@@ -2,9 +2,108 @@
  * \file Library with public methods provided by component openshift4-operators.
  */
 
+local com = import 'lib/commodore.libjsonnet';
 local kap = import 'lib/kapitan.libjsonnet';
 local kube = import 'lib/kube.libjsonnet';
 
-// Export library functions here
+local inv = kap.inventory();
+
+local params = inv.parameters.openshift4_operators;
+local instanceParams(instance) =
+  local ikey = std.strReplace(instance, '-', '_');
+  params + com.getValueOrDefault(inv.parameters, ikey, {});
+
+local apigroup = 'operators.coreos.com';
+
+local validateInstance(instance, checkTargets=false, checkSource='') =
+  local supported_instances = std.set([
+    'openshift-operators',
+    'openshift-operators-redhat',
+  ]);
+
+  assert
+    std.setMember(instance, supported_instances) :
+    "\n  Invalid instance '%s' for component openshift4-operators." % [
+      instance,
+    ] +
+    '\n  Supported instances are %s' % [
+      supported_instances,
+    ];
+
+  local appentry = 'openshift4-operators as %s' % [ instance ];
+  assert
+    !checkTargets || std.member(inv.applications, appentry) :
+    "\n  Unknown openshift4-operators instance '%s' for %s." % [
+      instance,
+      checkSource,
+    ] +
+    "\n  Did you forget to configure application '%s'?" % [
+      appentry,
+    ];
+
+  instance;
+
+
+local Subscription(name) =
+  kube._Object(apigroup + '/v1alpha1', 'Subscription', name) {
+    spec: {
+      name: name,
+    },
+  };
+
+local OperatorGroup(name) =
+  kube._Object(apigroup + '/v1', 'OperatorGroup', name);
+
+local globalSubscription =
+  function(
+    instance,
+    name,
+    channel,
+    source=instanceParams(instance).defaultSource,
+    sourceNamespace=instanceParams(instance).defaultSourceNamespace,
+    installPlanApproval=instanceParams(instance).defaultInstallPlanApproval
+  )
+    local _instance = validateInstance(
+      instance,
+      checkTargets=true,
+      checkSource='subscription %s' % [ name ]
+    );
+    Subscription(name) {
+      metadata+: {
+        namespace: _instance,
+      },
+      spec+: {
+        channel: channel,
+        installPlanApproval: installPlanApproval,
+        source: source,
+        sourceNamespace: sourceNamespace,
```

Comment:
`startingCSV` should be optional configurable.

Reply:
I don't see `startingCSV` documented anywhere. Please note that you can still customize your subscriptions after creating them with the library as follows:

```
operatorlib.namespacedSubscription('openshift-operators', 'my-subscription', 'my-channel') { 
  spec+: { startingCSV: 'my-starting-csv' }
}
```

Reply:
```
$ oc explain sub.spec
KIND:     Subscription
VERSION:  operators.coreos.com/v1alpha1

RESOURCE: spec <Object>

DESCRIPTION:
     SubscriptionSpec defines an Application that can be installed

FIELDS:
   channel	<string>

   config	<Object>
     SubscriptionConfig contains configuration specified for a subscription.

   installPlanApproval	<string>
     Approval is the user approval policy for an InstallPlan. It must be one of
     "Automatic" or "Manual".

   name	<string> -required-

   source	<string> -required-

   sourceNamespace	<string> -required-

   startingCSV	<string>
```
A example with a specified startingCSV - to pin the version - is:
```
apiVersion: operators.coreos.com/v1alpha1
kind: Subscription
metadata:
  name: strimzi-kafka-operator
  namespace: openshift-operators
spec:
  channel: stable
  installPlanApproval: Manual
  name: strimzi-kafka-operator
  source: community-operators
  sourceNamespace: openshift-marketplace
  startingCSV: strimzi-cluster-operator.v0.21.1
```
However as long as it can be overridden, this should be fine.

Reply:
I added another documentation page explaining how the library functions work.